src/main/kotlin/cacophony/grammars/Analysis.kt

```diff
@@ -1,44 +1,169 @@
+    private val symbolToStateReference = stateReferenceToSymbol.map { (u, v) -> Pair(v, u) }.toMap()
+    val firstMap: Map<DFAStateReference<StateType, SymbolType, ResultType>, Set<SymbolType>>
+
+    private fun fromState(ref: DFAStateReference<StateType, SymbolType, ResultType>): Entity<StateType, SymbolType, ResultType> {
+        val sym: SymbolType? = stateReferenceToSymbol[ref]
+        return if (sym == null) {
+            Entity.Ref(ref)
+        } else {
+            Entity.Sym(sym)
+        }
+    }
+
+    private fun fromSymbol(sym: SymbolType): Entity<StateType, SymbolType, ResultType> {
+        return Entity.Sym(sym)
+    }
+
+    private fun toState(entity: Entity<StateType, SymbolType, ResultType>): DFAStateReference<StateType, SymbolType, ResultType>? {
+        if (entity is Entity.Ref) {
+            return entity.ref
+        }
+        if (entity is Entity.Sym) {
+            return symbolToStateReference[entity.sym]
+        }
+        return null
+    }
+
+    private fun toSymbol(entity: Entity<StateType, SymbolType, ResultType>): SymbolType? {
+        if (entity is Entity.Ref) {
+            return stateReferenceToSymbol[entity.ref]
+        }
+        if (entity is Entity.Sym) {
+            return entity.sym
+        }
+        return null
+    }
+
+    init {
+        val nullableSymbols = nullable.mapNotNull { stateReferenceToSymbol[it] }.toSet()
+        val firstGraph = mutableMapOf<Entity<StateType, SymbolType, ResultType>, MutableSet<Entity<StateType, SymbolType, ResultType>>>()
+        for (dfa in automata.values) {
+            for (state in dfa.getAllStates())
+                firstGraph.computeIfAbsent(fromState(Pair(state, dfa))) { mutableSetOf() }
+            for (production in dfa.getProductions()) {
+                val from = fromState(Pair(production.key.first, dfa))
+                val to = fromState(Pair(production.value, dfa))
+                val firstSet = firstGraph.getOrPut(from) { mutableSetOf() }
+                firstSet.add(fromSymbol(production.key.second))
+                if (nullableSymbols.contains(production.key.second)) {
+                    firstSet.add(to)
+                }
+            }
+        }
+        for (key in firstGraph.keys)
+            println(key.toString() + " " + firstGraph[key].toString())
+        val closure = transitiveClosure(firstGraph)
+        for (key in closure.keys)
+            println(key.toString() + " " + closure[key].toString())
+        firstMap =
+            closure.mapNotNull { (key, values) ->
+                val newKey = toState(key) ?: return@mapNotNull null
+                val newValues = values.mapNotNull { toSymbol(it) }.toSet()
+                return@mapNotNull Pair(newKey, newValues)
+            }.toMap()
+    }
+}
+
+fun <VType> transitiveClosure(graph: Map<VType, Collection<VType>>): Map<VType, Set<VType>> {
```

Comment:
you can extract `transitiveClosure` and `findStronglyConnectedComponent` somewhere, they seem to be generic functions on graphs that could potentially be reused

and even if they won't be reused, you specifically made them generic and they just don't belong in this file

Reply:
Created GraphUtils in utils package. I also moved getReachableFrom there.